Initial

Mottie opened the pull request:
Adding a basic structure for the node module.

julmot commented:
Thanks for spending time on this module!

First of all, let's focus on the spec (README.md):

- `diacritics.setVersion("v1");` makes sense, to explicitly define which version of the API should be used. But, is this necessary? Imagine the following case:

    I'm a user that downloads the module. In my `package.json` the current version is defined (at the time of the installation). This version internally uses the `v1` API, as it's the most recent at the time of the module version. After a time a new API version is released, `v2`. Then this module still uses `v1` successfully. And only after a new version of the module is installed/updated, the `v2` API is being used. This means that a user would implicitly specify the API version by the module version. I think that is agreeable. If you have arguments against, I think it would also be okay to let this option exist and only specify a default value, e.g. `v1` that will be updated with new releases (in case there is a new API version).

- `Parsed values must be greater than zero and less than the current major release version.` Or equal, am I right?

- In all response values included in the docs, is is possible that you've forgot the language variants ("de" => "de" instead of "de")?

```javascript
{
    "de": {
        "metadata": { ... }
        "data": { ... }
    }
}
```

- I'm missing the "build" things. Having functions to transliterate strings is important too, but first of all, users should be able to generate an array/object of diacritics into their scripts using this module. The API needs to offer routes for this first, I know, but we can neverthless document or discuss how to achieve that
- What I'm missing at the current set of methods is the way to filter by multiple filters. The API can handle multiple parameters, the module doesn't. I imagined a placeholder in files including a URI of the API, this would be easy to implement and would allow `&&` filters

Independently of that, what I've asked myself today is: Why didn't we include countries in the metadata of the database, just continents?

Mottie commented:
> To explicitly define which version of the API should be used. But, is this necessary? 

You make a valid point... the module should be getting the version from the module package.json and not the package.json at GitHub.

I'm not sure if it is necessary, but I don't think it would hurt anything to include it.

> Or equal, am I right?

Yep, I need to add an "equal to" in there.

> In all response values included in the docs, is is possible that you've forgot the language variants ("de" => "de" instead of "de")?

Nope, in the module, I explicitly flattened the resulting object... I was thinking it would be easier to iterate through a list of variants than an object with language > variants. Do you think the outer wrapper is necessary in this case?

> users should be able to generate an array/object of diacritics into their scripts using this module

Yeah, I haven't gotten to that part yet... the `getDiacritics` function was the first step down that road... it accepts a string, and not just one character.

> I imagined a placeholder in files including a URI of the API

I started to add a `diacritics.replacePlaceholder` entry to the readme, but I got stuck...

```
#### `diacritics.replacePlaceholder(string, options)`

 Replaces placeholders within the string with the targeted diacritic values

- The `string` parameter contains text and/or HTML with a diacritic placeholder value(s) to
  be replaced.
- The `options` parameter contains the following settings:
  - `placeholder` - format defaults to `<% diacritics: {url} %>`.
  - `output` - defaults to `decompose` (values set as `decompose` or `base`).
  - `variant` - variant to use; if not defined, the first variant result will be used.
  - `callback` - function to allow customization: `function(data) { return data; }`
```

What I'm stuck on is how to handle variants.

> Why didn't we include countries in the metadata

Is that really necessary? The list of countries that speak a language would be very extensive since just about every country speaks many languages, even if only in a small community. What would be our cutoff? Or are you only referring to "official" languages of a country?

julmot commented:
> You make a valid point... the module should be getting the version from the module package.json and not the package.json at GitHub.

We've misunderstood here. What I meant was that `diacritics.setVersion("v1");` should not be _required_. If I'm calling a method without setting a version before, a default version of the API should be used. But this version can't be the same as the npm module version because they are two completely different things (we may update the npm module because of new features for example, but that doesn't mean that we need to update the API version too).  
Therefore, I meant that we have an internal default version e.g.

```javascript
class XYZ{
    constructor(){
        this.apiVersion = "v1";
    }
}
```

This default version may be updated on new npm module releases. 

> Do you think the outer wrapper is necessary in this case?

I think so. Imagine if we would have French Spanish (fr => es) and also normal Spanish (es => es), then there would be two equivalent keys (es).

> What would be our cutoff? Or are you only referring to "official" languages of a country?

Yeah, definitely just official languages.

> I started to add a diacritics.replacePlaceholder entry to the readme, but I got stuck...

Let me come back to this with more time.

Mottie commented:
> But this version can't be the same as the npm module version because they are two completely different things

Doh, my mistake. If we do not allow the user to choose the database version, how will they be able to acces a different database version? Will we need to create separate npm modules to access each?

> Imagine if we would have French Spanish...

Hmm, ok... I'll need work on it.

> Yeah, definitely just official languages.

Alright, are you going to update the database? Or would you like me to do it?

> diacritics.replacePlaceholder

I've updated the spec in the readme a bit more... please let me know what you think.

julmot commented:
> how will they be able to acces a different database version?

Let's go back a step: Why do they need to know anything about database versions? We've introduced database versioning and later API versioning in case there are structural JSON changes from time to time. Previously I thought the npm module includes only functions that actually generate array/objects/RegExp/... in JavaScript, but not the raw data. In this case they never need to know anything about a version, just the npm module itself needs to know where it should fetch the data from.  
Now, since you've introduced several functions that output the original JSON from the API, the version becomes important for them too. But, with the function `setVersion` they are still be able to overwrite the default version, aren't they? Everything should be fine.

> Will we need to create separate npm modules to access each?

No user needs to _create_ our npm module, just install it. In case the installed npm module is targeting to an old API version then they either need to run `$ npm update` as we'll update the npm module internally pointing to the new API version when there's one. Or they use the `setVersion` method to overwrite the default version.

> Alright, are you going to update the database? Or would you like me to do it?

Let's add this to the backlog ([card](https://github.com/diacritics/database/projects/1#card-1071093)) and work on it when the npm module is worked out.

> I've updated the spec in the readme a bit more... please let me know what you think.

I'll come back to this asap.

Mottie commented:
>

Actually, now that I think about it... the variant naming is not explicitly defined in the database spec... I think they need to defined using the [IETF language tags](http://data.okfn.org/data/core/language-codes#resource-ietf-language-tags). Doing this would prevent the conflict you mention and I think make the resulting data from this API a bit cleaner to use.

I am having trouble finding the actual current definitions for IETF tags, but I did find this W3C standard (https://www.w3.org/International/articles/language-tags/) that pretty much says the same thing in a blue comment box

> RFCs are what the IETF calls its specifications. Each RFC has a unique number. Unfortunately, it is not possible to tell, when reading RFC 1766 or RFC 3066 that these specifications have been obsoleted and replaced by other specifications.

Mottie commented:
Derp, the W3C page links to this definition: http://www.rfc-editor.org/rfc/rfc5646.txt

Mottie commented:
According to that spec, we would also be able to add a "catch-all" diacritic section using a `x-fallback` (or `x-` whatever) variant for diacritics that we can not find an associated language or variant. So that would avoid the [`en` variant addition that I previously proposed](https://github.com/diacritics/database/issues/1#issuecomment-264923329).

Mottie commented:
Ok, I only have the `replacePlaceholder` function left to work on. Can you think of anything other functions that might be nice to include?

julmot commented:
> Can you think of any other functions that might be nice to include?

I'll look at your work tonight.

Mottie commented:
I should have the `replacePlaceholder` function done sometime today.

julmot commented on .travis.yml:
```diff
@@ -0,0 +1,3 @@
+language: node_js
+node_js:
+  - "stable"
```
You need to have a `script` in order to properly use Travis CI later. Add the command `npm run linting` and `npm test` and take [this file](https://github.com/julmot/mark.js/blob/master/.travis.yml) as an example.

Mottie replied:
Done

julmot commented on README.md:
````diff
@@ -0,0 +1,713 @@
+
+# Installation
+
+```console
````
Is `console` valid? I've always used `bash` for things like that.

Mottie replied:
LOL I think I copied this from the API readme....

julmot replied:
Where?

Mottie replied:
LOL, I can't find it... I'll change it to `bash`.

Mottie replied:
Done

julmot commented on README.md:
````diff
@@ -0,0 +1,713 @@
+
+```js
+var diacritics = require("diacritics-transliterator");
+diacritics.setVersion("v1");
````
I'd vote to remove this from the default example, as we've discussed to make this optionally.

Mottie replied:
Remove the entire entry? Or just the example?

julmot replied:
I thought `diacritics.setVersion("v1");` is optional, so we don't need to have this explicit named in the minimal example.

Mottie replied:
Example removed.

julmot commented on README.md:
````diff
@@ -0,0 +1,713 @@
+
+Each API method requires the `diacritics-transliterator` module.
+
+```js
````
Is `js` valid? I've always used `javascript`

Mottie replied:
Yes, `js` is valid.

julmot replied:
Do you have a resource?

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+
+The `version` parameter (`string` or `number`) will accept the following values:
+
+- `v#`, where `v` is the abbreviation for "version" and `1` is the major release value.
```
I'm confused, do you mean `#` instead of `1` here?

Mottie replied:
Umm, yes? LOL

<img title=":_ tuzki_dizzy :" src="https://cloud.githubusercontent.com/assets/136959/13203535/61480e08-d880-11e5-9ea8-7fc03c7d7b56.gif">

Moot point if you think the entire entry needs to be removed.

julmot replied:
No, I wans't talking about removing the entire `setVersion` method.

julmot commented on README.md:
````diff
@@ -0,0 +1,713 @@
+var diacritics = require("diacritics-transliterator");
+var test = diacritics.getVariant("test");
+// test => { "message": "Variant 'test' was not found" }
+```
````
In case of `language` you've specified `ISO 639-1`, but in a standard is missing for `variant`.

Mottie replied:
Done

julmot commented on README.md:
````diff
@@ -0,0 +1,713 @@
+var diacritics = require("diacritics-transliterator");
+var string = diacritics.formatUnicode("T\\u00E9st");
+// string => "Tést"
+```
````
Why do we offer such method? 

Btw.: Maybe it's helpful to use `String.raw` here instead of double-escaping the string.

Mottie replied:
It is needed because the database stores escaped unicode values (e.g. `\\u00E9`). Storing `\u00E9` would be the same as providing the `raw` value.

I'm not sure what you mean by using `String.raw`, but this method converts the double escaped unicode into a raw unicode value.

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+
+Returns the base data for the matching base value within the array.
+
+The `array` parameter contains either an `array` of diacritic base string, or a `string` containing a single base (internally converted into an array), of characters to process.
```
`(internally converted into an array)` not necessary to mention (gernally).

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+Replaces diacritics within the string with either the base or decomposed value.
+
+- The `string` parameter (type: `string`) will replace any diacritic characters with base or decomposed values. Any non-diacritics characters will not be modified.
+- The `type` parameter (type: `string`) is set to either "base" or "decompose". Defaults to "base". If a "base" value is not found, a "decompose" value will be used instead; and vice-versa.
```
I don't understand this sentence: `If a "base" value is not found, a "decompose" value will be used instead; and vice-versa.`. So, what's the fallback, `base` or `decompose`?

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+// The base of ß is still ß
+var generic = transliterate("¿abcñ-ß123?"); // "?abcn-ß123?"
+// If no diacritics match the set language variant, the original string is returned
+var unchanged = transliterate("¿abcñ-ß123?", "base", "test"); // "¿abcñ-ß123?"
```
General question: How are exceptions handled? I mean "test" is invalid, you could also expect an error. The same applies to an invalid value instead of "base" or "decompose".

Mottie replied:
Invalid `type` and `variant` will return the original string.

How should we handle exceptions? Add a console log error?

julmot replied:
What about throwing errors, just an idea? For example: `throw new Error("Invalid 'type'");`

Mottie replied:
The `transliterate` now throws an error when the `type` is invalid, but still ignores invalid or missing variants.

I was considering using [franc](https://github.com/wooorm/franc) or [languagedetect](https://github.com/FGRibreau/node-language-detect) to determine the language, then load the associated diacritcs, but I don't know how accurate that would be for language variants that might have different base and decompose values.

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+<table>
+  <thead>
+    <tr>
+      <th>Option
```
Is there a reason why you're not setting the closing tags?

Mottie replied:
Closing tags are optional... this makes it easier to read

ref: https://www.w3.org/TR/html5/syntax.html#optional-tags

Mottie replied:
😞

Mottie replied:
LOL, I've already added them <img title=":_bipolar:" src="https://cloud.githubusercontent.com/assets/136959/13512241/15d8362c-e15f-11e5-8c32-35b78deb9017.gif">

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+### diacritics.replacePlaceholder(string [, options])
+
+Replaces placeholder(s) within the string with the targeted diacritic values:
+- The `string` parameter (type: `string`)  contains text and/or HTML with a diacritic placeholder value(s) to be replaced.
```
I wouldn't explitly call it "HTML" as it may be "JS" too.

Mottie replied:
Oh, I meant the string may contain HTML tags.

julmot replied:
I know, but this confuses me a little bit. It may also be JS. In any case, it's text, so why is it necessary to explicit mention this?

Mottie replied:
I removed the `HTML` part and just left it as text... I added the note about HTML because the diacritic placeholder looks like HTML.

julmot replied:
You're right, that's the point I wanted to address. As I've mentioned in [my comment](https://github.com/diacritics/database/issues/1#issuecomment-262605304) if you have a placeholder like `<% diacritics: base = u; decompose %>` then this suits perfectly for HTML files and JS build templates. But what about JS files you want to use for development and you don't want to make it invalid with such a placeholder (files that work with the placeholder in it). In this case a placeholder like `/* diacritics: /?language=DE */` makes sense because it doesn't make the file invalid (because it's a comment). We need to have a way to support all these project variants.

Mottie replied:
I've included a placeholder option... you can modify it however you wish. Look at [this test](https://github.com/diacritics/node-diacritics-transliterator/commit/85e8cb6047ac371e7b50a71d66cb090aaa3abcf7#diff-65a9616c82f38fb3b6e9e6373a914566R537) which sets the diacritic to use `<!== values:{data} ==>`

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+  - Example: `<% diacritics: base=o;equivalents.unicode %>` will be replaced with `\u00FC,u\u0308,\u00FA,u\u0301` - this example is only showing the results from `de` and `es` languages; there will be a lot more once there is more data. The result can be reformatted using the `output` callback function.
+- The `options` parameter contains the following settings:
+  - `placeholder` - (type: `string`) format defaults to `<% diacritics: {query} %>`.
+  - `output` - (type: `function`) function to allow customization of the output string: `function(placeholder, result) { return result.join(""); }`
```
Please describe what this example function does.

Mottie replied:
This was just the preliminary outline... almost all of it has changed.

I want to push the replacePlaceholder code to this PR, but I'm not sure if I should reference `source` or `sources`, `country` or `countries`...

julmot replied:
Reference the new structure, I'll update the API as my next to-do (or check if it's necessary at all)

Mottie replied:
I'll wait until we merge the database PR... otherwise the tests will start failing.

julmot replied:
Sure!

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+  - The query string contains two parts: `type;data`
+  - The `type` will match the diacritics.io database queries (e.g. `base=o`)
+  - The `data` will target the data to include (e.g. `alphabet`, `native`, `base`, `decompose` or `equivalents.unicode`) - including `metadata` or `data` is not required.
+  - Example: `<% diacritics: base=o;equivalents.unicode %>` will be replaced with `\u00FC,u\u0308,\u00FA,u\u0301` - this example is only showing the results from `de` and `es` languages; there will be a lot more once there is more data. The result can be reformatted using the `output` callback function.
```
What are valid values other than `equivalents.unicode`? For example how to replace it with an array like [this](https://github.com/julmot/mark.js/blob/master/src/mark.js#L209)?

Mottie replied:
This has been changed, alot!

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+# Related
+
+- [diacritics database](https://github.com/diacritics/database).
+- [diacritics API](https://github.com/diacritics/api).
```
Please add just a single sentence what these projects are about.

julmot commented on README.md:
```diff
@@ -0,0 +1,713 @@
+- [diacritics database](https://github.com/diacritics/database).
+- [diacritics API](https://github.com/diacritics/api).
+
+# [License](LICENSE)
```
About the licensing in general:

1. We have new year in front of our doors, so I'd vote to update the "LICENSE" file to show Copyright ... 2017 or at least 2016-2017.
2. As the entire project is built as team work I'd vote to not use my name here as the copyright holder, also not in the other repositories. I'd call it something like "Copyright (c) 2017 The Diacritics Team". But, on the other hand I'm not sure if this is acceptable from a legal point of view (as it's not clear who is meant with this "team"). Needs some investigation first.   
Furthermore I'd not explicit name the license here (LICENSE file is enough imho).

julmot commented on index.js:
```diff
@@ -0,0 +1,749 @@
+"use strict";
+
+const databaseURL = "http://api.diacritics.io/",
+
```
Please remove the unnecessary line break

julmot commented on index.js:
```diff
@@ -0,0 +1,749 @@
+        current = major > 1 ? major : 1;
+    }
+    module.exports.currentVersion = "v" + current;
+    // make readonly?... modified in test.js for testing
```
Is this a debug comment?

Mottie replied:
It's a reminder since we were discussing removing the ability to set versions... I'll go ahead and remove it.

julmot commented on package.json:
```diff
@@ -0,0 +1,47 @@
+    "node": ">=4"
+  },
+  "devDependencies": {
+    "ava": "*",
```
As ava seems to not be in a stable relase, using `*` may break in future.

julmot commented on test/settings.json:
```diff
@@ -0,0 +1,6 @@
+{
```
What's this file about? Couldn't find something in the `CONTRIBUTING.md` file and this seems to only be used in test cases.

Mottie replied:
This settings file has settings to be used during testing.

Each method checks for a `module.exports.debug.{name}` and outputs a message to the console for debugging purposes.

I'll update the contributing file.

julmot commented:
Some general feedback:
- I haven't had the time to test the module functionality, I hope to have the time tomorrow
- I'm missing explicit documented return or parameter types/values. You're documenting this as a comment in an example or inside brackets (e.g. "(type: string)"). Also I'm missing an explicit documented fallback return value, e.g. `If a language is not found, an object with an error "message" is returned`, which means `If a language isn't found, an object with a property "message" is returned`.
- As you can see in [this card](https://github.com/diacritics/api/projects/1#card-919969) I need to add a way to allow filtering by e.g. one or multiple continents. The spec will then need an update too, because you could then theoretically use `diacritics.getContinent(["EU"])`. This applies also to `metadata.country`.
- I didn't fully understand `diacritics.createRegExp` with all its options. I'll try to understand this tomorrow again.

Mottie commented on README.md:
```diff
@@ -0,0 +1,713 @@
+  - `placeholder` - (type: `string`) format defaults to `<% diacritics: {query} %>`.
+  - `output` - (type: `function`) function to allow customization of the output string: `function(placeholder, result) { return result.join(""); }`
+
+The returned value will be a `string` (converted from an array by the `output` callback) containing unique diacritic that match the query string values - all duplicates are removed.
```
is that an em-dash or en-dash?

julmot replied:
You've used a hyphen, I've used a en dash.

Mottie replied:
LOL yeah, I don't have that on my keyboard, so I'd have to copy & paste it every time. I'm just not used to it 😉

julmot replied:
I don't too, so I'm on your site. You can use ALT + 0150 to generate it.

Mottie commented:
Ok, I've worked on almost all of your suggestions. I also added the `replacePlaceholder` method. It will work with the database still set to `source` and `countries`. I'll update it when that PR goes live.

The only thing I still haven't had a chance to do is update the documentation to include returned values.... too tired now. I'll work on it tomorrow.

Mottie commented:
Oh, and I've already started working on a CLI... I should have that finished by the weekend.

julmot commented on package.json:
```diff
@@ -0,0 +1,47 @@
+  "main": "index.js",
+  "scripts": {
+    "test": "ava test/*.js",
+    "linting": "./node_modules/.bin/eslint './*.js'"
```
The path doesn't seem to be necessary. `"linting": "eslint './*.js'"` seems to be enough, like you've done this with `ava`. It's documented [here](https://docs.npmjs.com/misc/scripts#path). I didn't know this too.

julmot commented on CONTRIBUTING.md:
````diff
@@ -18,4 +18,54 @@ After you have contributed something check your code by using:
+diacritics.getVariant("de");
+// logged as: "http://api.diacritics.io/v1/?language=de; {loaded from cache}"
+```
+
````
I guess that this code is from a users point of view as we don't require the module in `index.js`. If so, then the debug option is missing in the README.md

Mottie replied:
I added the debugging while I was adding tests because I wanted to see why some tests were failing. I don't think it would be necessary to have the output as a user, so I didn't add it as an option to each method.

The `debug.regexpTests` setting was originally included with the `debug.regexp`. When set to `true`, all you would see was screen after screen of logs, so I split it to two parts. This setting isn't really necessary. I can remove it in the final version.

julmot replied:
If it's just for internal usage I think it shouldn't be exported for public usage.

julmot commented:
@Mottie Yesterday I've forgot to mention: WOW. Incredibly huge progress! Feels like there's light at the end of the tunnel. :tada:

Anyway, unfortunately I'm still not finished with the review. I guess it will take some more days as there's so much to test. I've not tested one single line from a functional point of view, nor have I started reviewing the code (test files as well as `index.js`).

But, I've done a quick scan of the `index.js` and I think it would help a lot to reduce the code complexity and documentation structure first. 

### Modularity

Is there a chance to get it more modular? With modular I mean smaller functions, maybe splitted into different classes if necessary. Especially the following functions are very long and not reusable:

- `validatePath`
- `extractPlaceholderSettings`
- `module.exports.createRegExp`
- `followPath`
- `module.exports.replacePlaceholder`

I personally use my display as an indication if a function is too long/complex. If it's not possible to view the function (including comments, but not the JSDOC function header) on my screen (24"), then this tells me that I need to refactor it into smaller parts (of course there are other factors like the font size, etc.). This makes it easy to read and understand as well as open for reuse.  

### Redundancy

The other thing I've seen is redundant code. I guess this would be solved with a modular refactoring, but just want to mention it for the sake of completeness. For example I've noticed that you are checking for a cached URL four times:

```javascript
if(module.exports.cache[url]) {
    // get cached variant
    data = module.exports.cache[url];
    if(module.exports.debug.server) {
        console.log(url + "; {loaded from cache}");
    }
}
```

You could for example just build a `Cache` class or method that handles this once.

### module.exports vs function

Currently there are static functions at the top of the file and `module.exports` functions at the bottom. Both of them are treated like normal functions, containing logic. This means you need to write `module.exports.X` all the time.  
I personally find this annoying and I think it would be better to have function/variable names and just export them using `module.exports`, not storing much logic within these exports.

### Consistent Documentation Structure

I'd prefer to have a consistent API documentation structure. The documentation currently contains all the information, but inside sentences. I'd like to orient on professional technical documentations which would for example be something like this:

```bash
### diacritics.setVersion()

Upon initialization, this module is set to the latest major release version of the diacritics API. To change the version, use this function to set the API to the desired major release.  
{Add more documentation things here}

**Syntax**

const variant = instance.setVersion(version);

**Return Value**

The specified version code.

**Parameters**

<dl>
 <dt><code>version</code></dt>
 <dd>The version to set</dd>
</dl>

```

### Improvement Suggestion

First of all, it would help if you would move the code into a `./src/` folder and bundle related function into a class. Then create separate files when necessary.

If I have a look at the existing methods I associate a "filter" with all methods starting with "get*". Thus I would group them in a `Filter` class. Methods under "transliteration" would be grouped in a `Transliterator` or `Transliteration` class. All other seem to be common methods, therefore would be grouped in a `Core` class.

The `Core` class would be the parent of class `Filter` and `Transliterator.`


![transliterator](https://cloud.githubusercontent.com/assets/8310677/21555584/a73102f0-ce19-11e6-8c78-50b1bea0ea87.png)

This would be modular. If the `Filter` class would get too large, it would be possible to create subclasses for `metadata` and `data` too, so that `Filter` would be the parent class of e.g. `FilterMetadata` and `FilterData`.

### Conclusion

Anway, this is just my personal cristitm. You are not obliged to hold on to it. But, of course I'm convinced that realizing modular and reusable code is a good practice, otherwise I wouldn't recommend it. Anyway, I'm not sure if you expected such a review :laughing: But I personally would love if someone reviews my code and gives helpful feedback how to improve it. I hope you don't take this against me.

Thanks for your hard work!


Btw: I'.ve activated Codacy for the initial branch [here](https://www.codacy.com/app/diacritics/node-diacritics-transliterator/dashboard?bid=4038407).

I've added some more ESLint rules that will make the current code fail, just in case you'd like to spend some more time on a refactoring.

Mottie commented:
Hmm, only the cli tests are failing... `Error: spawn EACCES`? I googled the error and it looks like one solution is to use `sudo`?... I don't know much about linux, so I'm not sure how to add the command and still have the cli tests work on other operating systems.

julmot commented on cli.js:
```diff
@@ -0,0 +1,142 @@
+/*!***************************************************
+ * node-diacritics-transliterator cli
+ * Diacritic transliteration tools using diacritics.io
+ * API from http://diacritics.io/
```
API from http://api.diacritics.io/

julmot commented on cli.js:
```diff
@@ -0,0 +1,142 @@
+    input = (cli.input || []).join(" "),
+    flags = cli.flags,
+    xref = {
+        fu: flags.formatUnicode      || flags.fu || false ? "fu" : "",
```
Won't `|| false` be always false?

Mottie replied:
Derp!

julmot commented on README.md:
```diff
@@ -0,0 +1,1075 @@
+### getData
+
+Function to return specificially filtered data from the diacritics database. Include one or more filters to narrow the query. Invalid filters are ignored.
+
```
It does support filtering by an array [now](https://github.com/diacritics/api/commit/5bcc0c50cdb5189e4ed154d0ebe8b40409811bf9).

julmot commented on README.md:
````diff
@@ -0,0 +1,1075 @@
+
+```js
+const diacritics = require("diacritics-transliterator");
+/* showing all filters (currently only one setting per filter allowed) */
````
The examples help a lot!

julmot commented on README.md:
```diff
@@ -0,0 +1,1075 @@
+# node-diacritics-transliterator
+
+> Diacritic transliteration tools using diacritics.io API
```
I think that this is the most important sentence in the whole documentation and we need to elaborate a bit more what a user can do with this module. I think he won't read through the documentation to find out what it's about, so we should clarify this directly at the beginning to win him as a user. What do you think?

julmot commented on README.md:
```diff
@@ -0,0 +1,1075 @@
+
+If you contribute to this project, you are implicitly allowing your code to be distributed under [this license][license]. You are also implicitly verifying that all code is your original work.
+
+[license]: https://git.io/v1EBe
```
Fuck, man, you're a fucking writing machine 😆 I have flown over it and it looks good to me. I'll test how the documentation feels from a user point of view as soon as I migrate the diacritics module with mark.js.

But, the CLI documentation is missing completely. There's no need to document the CLI methods redundant, but at least specify that there's a CLI and how to see the available commands.

And how to test the CLI in development without installing it from npm?

julmot commented on index.js:
```diff
@@ -0,0 +1,25 @@
+module.exports.transliterate = require("./src/transliterate").transliterate;
+module.exports.createRegExp = require("./src/regexp").createRegExp;
+module.exports.replacePlaceholder =
+    require("./src/placeholder").replacePlaceholder;
```
Please `require` at the top.

julmot commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,21 @@
+After you have contributed something check your code by using:
+
+```bash
+$ npm run linting | npm test
````
That's not valid. Please use `&&`

julmot commented on README.md:
```diff
@@ -0,0 +1,1186 @@
+
+<dl>
+  <dt><code>version</code> (optional)</dt>
+  <dd>The version of the database API to use. Defaults to the current version.</dd>
```
`int` or `string`?

julmot replied:
Where to pass the version? `require("diacritics-transliterator")(version)`?

Mottie replied:
Oh, there aren't any parameters. I need to remove that. You set the version like this:

```js
const diacritics = require("diacritics-transliterator");
diacritics.version = "v1";
```

You can find this under basic functions > setVersion

julmot commented:
Here are some review notes regarding the node module. I'll test the CLI in a few minutes:

1.  When you're offline there should be a message that the module requires an internet connection. At a later point we should think about offline fallbacks.

2. Using the [basic example](https://github.com/diacritics/node-diacritics-transliterator/blob/initial/README.md#usage) and all further methods will display a warning:
![2017-08-25 09_40_18-cmd](https://user-images.githubusercontent.com/8310677/29704438-1c38c9c2-897a-11e7-9297-ed3d3b6d89ac.png)

3. Using `.getVersion()` on the required instance like described in the docs doesn't work for me:
![2017-08-25 09_43_22-cmd](https://user-images.githubusercontent.com/8310677/29704452-2b5c62ce-897a-11e7-942b-e5811f2730f9.png)
I guess it should be `.version` instead, like for the setting version.

4. The headlines `setVersion` and `getVersion` indicate functions, but are just one property. Please fix the contents afterwards.

5. Please document that `getData` is the main method for retrieving information

6. Regarding the following sentence in `getData`:

> The API supports setting filters as arrays for metadata only

According to [the docs](https://github.com/diacritics/api/tree/master/spec/v1) it currently only supports multiple values for fields where you can author multiple values, which means `country` and `continent`. However, I'll try to find a solution to allow multiple values generally.

7. I don't understand this sentence:

> If a single filter is used and set as an array, the returned object will contain a combination of all resulting metadata & data.

According to my tests, specifying `continent: "EU"` and `continent: ["EU"]` are returning the same output.

8. According to this sentence:

> All parameters are optional, but at least one must be set.

Does this mean that it wouldn't be possible to get all diacritics without any filter? That should be possible.

9. The example `language": ["de", "es"]` won't work currently, as mentioned above. However, I'll try to make it possible to allow array values generally, so no change is necessary here at the moment.

10. Is it true, that the method `getDiacritics` is a redundant subset method for only filtering by diacritics, while the `getData` method is more general? In this case we should remove it. If the argument to build this method was to allow filtering by multiple diacritics – as mentioned – I'll try to find a way to specify array values in general in the API.

11. In the `type` parameter of the transliterate method, please document what base and decompose properties are or add a reference to the corresponding database documentation section.

12. In this example a comma is missing: `const german = transliterate("¿abcñ-ß123?", "decompose" "de");`

13. Please name the `variant` of the transliterate method `languageVariant` to indicate that it's language-specific

14. If I understand the `createRegExp` correctly, the meaning is that you can create a regular expression based on a string that will then also e.g. catch all diacritics in the expression. So when I call it with `Test` I'd expect to also match diacritics with the base `e`. But that's not the case. Can you please describe the meaning behind this function in more detail especially for what it can be used?

15. A comma is missing after this: `placeholder: "<!-- diacritics: {data} -->"`

16. I've tried `replacePlaceholder`'s first example and it failed:

![2017-08-25 11_22_20-cmd](https://user-images.githubusercontent.com/8310677/29708067-bd5e6390-8987-11e7-9573-36c01bffad76.png)

The second example fails too:

![2017-08-25 11_23_20-cmd](https://user-images.githubusercontent.com/8310677/29708094-d6340cc6-8987-11e7-806d-bd7bb5ea432d.png)

Stopped testing this method at this point.

17. Typo `placeholderes`

18. Honestly, the `replacePlaceholder` documentation is way too complex. I don't understand the syntax or how I would integrate it. I think the examples are not practice-oriented. What I would like to know here is how do I generate an array [like this](https://github.com/julmot/mark.js/blob/8.11.0/src/mark.js#L302)? The methods above are targeting Node.js users that would transliterate strings at run-time. However, for web-based applications we need to generate an array like mentioned above to make it possible to generate the regular expression at runetime too, without calling the API at this time.

julmot commented:
Here are my review notes for the CLI:

1. Please make sure to have a consistent `--` or `-` prefix structure. For example on a global basis you're always using `--` (` --fu, --formatUnicode`) but inside the modules you're using both, `-` and `--` (`-t, --type`). I'd suggest to transform e.g. `--fu` into `-fu` too

2. Please rename `-v` to `-lv` to indicate that it's a language variant (as mentioned in the module review)

3. Why doesn't `diacritics "heey höö" --transliterate` transliaterate the given string? Only when I define `-v de` it seems to work. I'd expect that it transliterates the string including all language variants. As this would work in the module via Node.js itself:

```javascript
const transliterate = require("./index.js").transliterate;
const german = transliterate("töff");
// => toff
```

Mottie commented:
Wow, I have a lot of stuff to fix... sorry, I've been neglecting this module. It's been a busy few weeks for me. I'll try to get to this over this weekend.

julmot commented:
fyi: It should now be possible to allow specifying multiple values per parameter generally, e.g.:

http://api.diacritics.io/v1/?language=de,es,tr&diacritic=%C3%BC,%C3%B6,%C3%A4&decompose=oe

I'd appreciate if you'd test this too. If you're noticing any exception, please let me know.

Mottie commented:
> 1. When you're offline there should be a message that the module requires an internet connection.

We would need to include [is-online](https://github.com/sindresorhus/is-online) as a dependency. Or do you know of a simpler solution?

> 2. Using the basic example and all further methods will display a warning:

Yeah, that's because the module isn't using asynchronous requests. I mentioned this before and asked if you could look at it.

> 3 & 4... `.getVersion()`

Fixed in docs

> 5, 6 & 9... `.getData()` & "I'll try to find a solution to allow multiple values generally"

* Added a statement about it being the main function
* I haven't looked at the code in so long... I'm having a hard time figuring out why I did some things. I'm sure the code could use some cleaning up. I'll wait until you get a chance to look at the database and update the methods.

> 7. `continent: "EU"` and `continent: ["EU"]` are returning the same output

The statement "If a single filter is used and set as an array, the returned object will contain a combination of all resulting metadata & data" means if (with the current database) you set continent as `["AN", "AS"]` it will include diacritics from both continents. You will get a different result if you replace "AN" with "EU".

> 8... get all diacritics without any filter? That should be possible.

Yes, I believe that was the way I set it up (it's been a while since I looked at this code)... I'll dig into the code again soon.

>10... method `getDiacritics` is a redundant

It isn't redundant... `getData` only passes the query to the database whereas the `getDiacritics` function extracts out any diacritics from the string containing both diacritics and non-diacritics that you pass to it. This is why [`regenerate`](https://www.npmjs.com/package/regenerate) was included as a dependency.

> 11... document what base and decompose properties

Added links to the database spec.

> 13...  transliterate method `languageVariant`

Done; but, I'm not sure why a bunch of tests are failing now, but I at least wanted to start fixing the readme issues...

> 15 & 17.... comma is missing & Typo `placeholderes`

Fixed

----

I'll keep working on this, but a quick note about the CLI... the single dash (`-`) only works with a single letter parameter, and a double dash (`--`) works for one or more letters in a parameter, so setting `-lv` won't work... I'll look at the code and make everything consistent.

julmot commented:
Hey mate,

Is this PR ready for another review?

Mottie commented:
No sorry, it still needs some work... I've been really busy lately!

julmot commented:
fyi: I've adjusted the [initial milestone](https://github.com/diacritics/node-diacritics-transliterator/milestones) due date to 31 October '17. Hopefully we can finalize it until then. But don't feel oppressive. If you don't have time for it currently, it's fine too.

I have already one extension in mind but let's finish this first.

Mottie commented:
> due date to 31 October '17

Ok, that sounds reasonable. I'll move this up on my priority list.

julmot commented:
Since I've now started implementing a minimum viable product (which is not completed yet) I'm going to close this huge PR and we'll implement parts of it from time to time.
